superset-frontend/src/dashboard/actions/sliceEntities.ts

```diff
@@ -0,0 +1,178 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import rison from 'rison';
+import {
+  DatasourceType,
+  FeatureFlag,
+  SupersetClient,
+  t,
+} from '@superset-ui/core';
+import { addDangerToast } from 'src/components/MessageToasts/actions';
+import { getClientErrorObject } from 'src/utils/getClientErrorObject';
+import { isFeatureEnabled } from 'src/featureFlags';
+import { Dispatch } from 'redux';
+import { Slice } from '../types';
+
+const FETCH_SLICES_PAGE_SIZE = 200;
+
+export function getDatasourceParameter(
+  datasourceId: number,
+  datasourceType: DatasourceType,
+) {
+  return `${datasourceId}__${datasourceType}`;
+}
+
+export const ADD_SLICES = 'ADD_SLICES';
+function addSlices(slices: { [id: number]: Slice }) {
+  return { type: ADD_SLICES, payload: { slices } };
+}
+
+export const SET_SLICES = 'SET_SLICES';
+function setSlices(slices: { [id: number]: Slice }) {
+  return { type: SET_SLICES, payload: { slices } };
+}
+
+export const FETCH_ALL_SLICES_STARTED = 'FETCH_ALL_SLICES_STARTED';
+function fetchAllSlicesStarted() {
+  return { type: FETCH_ALL_SLICES_STARTED };
+}
+
+export const FETCH_ALL_SLICES_FAILED = 'FETCH_ALL_SLICES_FAILED';
+function fetchAllSlicesFailed(error: string) {
+  return { type: FETCH_ALL_SLICES_FAILED, payload: { error } };
+}
+
+const parseResult = (result: any[]) => {
+  const slices: { [id: number]: Slice } = {};
+  result.forEach((slice: any) => {
+    let form_data = JSON.parse(slice.params);
+    form_data = {
+      ...form_data,
+      // force using datasource stored in relational table prop
+      datasource:
+        getDatasourceParameter(slice.datasource_id, slice.datasource_type) ||
+        form_data.datasource,
+    };
+    slices[slice.id] = {
+      slice_id: slice.id,
+      slice_url: slice.url,
+      slice_name: slice.slice_name,
+      form_data,
+      datasource_name: slice.datasource_name_text,
+      datasource_url: slice.datasource_url,
+      datasource_id: slice.datasource_id,
+      datasource_type: slice.datasource_type,
+      changed_on: new Date(slice.changed_on_utc).getTime(),
+      description: slice.description,
+      description_markdown: slice.description_markeddown,
+      viz_type: slice.viz_type,
+      modified: slice.changed_on_delta_humanized,
+      changed_on_humanized: slice.changed_on_delta_humanized,
+      thumbnail_url: slice.thumbnail_url,
+      owners: slice.owners,
+      created_by: slice.created_by,
+    };
+  });
+  return slices;
```

Comment:
this seems be worth replacing by `.reduce`

```suggestion
  return result.reduce((slices, slice: any) => ({
    ...slices,
    [slice.id]: {
      slice_id: slice.id,
      slice_url: slice.url,
      slice_name: slice.slice_name,
      form_data,
      datasource_name: slice.datasource_name_text,
      datasource_url: slice.datasource_url,
      datasource_id: slice.datasource_id,
      datasource_type: slice.datasource_type,
      changed_on: new Date(slice.changed_on_utc).getTime(),
      description: slice.description,
      description_markdown: slice.description_markeddown,
      viz_type: slice.viz_type,
      modified: slice.changed_on_delta_humanized,
      changed_on_humanized: slice.changed_on_delta_humanized,
      thumbnail_url: slice.thumbnail_url,
      owners: slice.owners,
      created_by: slice.created_by,
    },
  }, {}), 
```

Reply:
Done 👍🏼